File: HexEditor.java
// HexEditor.java
// Версия на Java с использованием NIO, MappedByteBuffer, MessageDigest

import java.io.*;
import java.nio.*;
import java.nio.channels.*;
import java.nio.file.*;
import java.security.*;
import java.util.zip.*;
import java.util.*;

public class HexEditor {
    private final Path path;
    private final FileChannel channel;
    private final MappedByteBuffer buffer;
    private final long size;
    private final int cols;

    public HexEditor(String filename, int cols) throws IOException {
        this.path = Paths.get(filename);
        this.cols = cols;
        this.channel = FileChannel.open(path, StandardOpenOption.READ, StandardOpenOption.WRITE);
        this.size = channel.size();
        this.buffer = channel.map(FileChannel.MapMode.READ_WRITE, 0, size);
    }

    public void view(long offset, long length) {
        if (length == 0) length = size - offset;
        if (offset < 0 || offset + length > size) {
            System.err.println("Смещение вне диапазона");
            return;
        }
        buffer.position((int) offset);
        byte[] data = new byte[(int) length];
        buffer.get(data);
        dump(data, offset);
    }

    private void dump(byte[] data, long baseOffset) {
        for (int i = 0; i < data.length; i += cols) {
            int end = Math.min(i + cols, data.length);
            long addr = baseOffset + i;
            System.out.printf("%s  ", colorize(String.format("%08X", addr), "\u001B[96m"));
            // HEX
            StringBuilder hex = new StringBuilder();
            for (int j = i; j < end; j++) {
                hex.append(String.format("%02X ", data[j]));
            }
            while (hex.length() < cols * 3) hex.append(' ');
            System.out.printf("%s  ", colorize(hex.toString(), "\u001B[92m"));
            // ASCII
            StringBuilder ascii = new StringBuilder();
            for (int j = i; j < end; j++) {
                byte b = data[j];
                ascii.append((b >= 32 && b < 127) ? (char) b : '.');
            }
            System.out.printf("%s\n", colorize(ascii.toString(), "\u001B[94m"));
        }
    }

    private String colorize(String text, String color) {
        return color + text + "\u001B[0m";
    }

    public void editByte(long offset, byte value) {
        if (offset < 0 || offset >= size) {
            System.err.println("Смещение вне диапазона");
            return;
        }
        buffer.put((int) offset, value);
    }

    public List<Long> findBytes(byte[] pattern, long start) {
        List<Long> positions = new ArrayList<>();
        if (start < 0) start = 0;
        if (start + pattern.length > size) return positions;
        for (long i = start; i <= size - pattern.length; i++) {
            boolean match = true;
            for (int j = 0; j < pattern.length; j++) {
                if (buffer.get((int) i + j) != pattern[j]) {
                    match = false;
                    break;
                }
            }
            if (match) positions.add(i);
        }
        return positions;
    }

    public int replaceBytes(byte[] pattern, byte[] replacement, long start) {
        if (pattern.length != replacement.length) {
            System.err.println("Длина шаблона и замены должны совпадать");
            return 0;
        }
        List<Long> positions = findBytes(pattern, start);
        for (long pos : positions) {
            for (int j = 0; j < replacement.length; j++) {
                buffer.put((int) pos + j, replacement[j]);
            }
        }
        return positions.size();
    }

    public long checksumCRC32() {
        CRC32 crc = new CRC32();
        crc.update(buffer);
        return crc.getValue();
    }

    public String checksumMD5() throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("MD5");
        md.update(buffer);
        return bytesToHex(md.digest());
    }

    public String checksumSHA1() throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("SHA-1");
        md.update(buffer);
        return bytesToHex(md.digest());
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.toString();
    }

    public void close() throws IOException {
        buffer.force();
        channel.close();
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Использование: java HexEditor <file> [команда] [опции]");
            return;
        }
        String filename = args[0];
        String command = "view";
        long offset = 0;
        int cols = 16;
        String bytesHex = null, replaceHex = null, checksumType = null, stringSearch = null, output = null;
        boolean edit = false;

        for (int i = 1; i < args.length; i++) {
            switch (args[i]) {
                case "-o": offset = Long.parseLong(args[++i]); break;
                case "--cols": cols = Integer.parseInt(args[++i]); break;
                case "-b": bytesHex = args[++i]; break;
                case "--replace-with": replaceHex = args[++i]; break;
                case "-s": stringSearch = args[++i]; break;
                case "-c": checksumType = args[++i]; break;
                case "--output": output = args[++i]; break;
                default:
                    if (args[i].matches("view|edit|find|replace|checksum")) {
                        command = args[i];
                    }
            }
        }

        HexEditor editor = new HexEditor(filename, cols);
        try {
            switch (command) {
                case "view":
                    editor.view(offset, 0);
                    break;
                case "edit":
                    if (bytesHex == null) {
                        System.err.println("Укажите --bytes");
                        return;
                    }
                    byte val = (byte) Integer.parseInt(bytesHex.trim(), 16);
                    editor.editByte(offset, val);
                    System.out.printf("Байт по смещению %d изменён на %02X\n", offset, val);
                    break;
                case "find":
                    byte[] pattern;
                    if (bytesHex != null) {
                        String[] tokens = bytesHex.split(" ");
                        pattern = new byte[tokens.length];
                        for (int i = 0; i < tokens.length; i++) {
                            pattern[i] = (byte) Integer.parseInt(tokens[i], 16);
                        }
                    } else if (stringSearch != null) {
                        pattern = stringSearch.getBytes();
                    } else {
                        System.err.println("Укажите --bytes или --string");
                        return;
                    }
                    List<Long> positions = editor.findBytes(pattern, offset);
                    if (!positions.isEmpty()) {
                        System.out.printf("Найдено вхождений: %d\n", positions.size());
                        for (long p : positions) System.out.printf("  %08X\n", p);
                    } else {
                        System.out.println("Не найдено");
                    }
                    break;
                case "replace":
                    if (bytesHex == null || replaceHex == null) {
                        System.err.println("Укажите --bytes и --replace-with");
                        return;
                    }
                    String[] patTokens = bytesHex.split(" ");
                    byte[] pat = new byte[patTokens.length];
                    for (int i = 0; i < patTokens.length; i++) pat[i] = (byte) Integer.parseInt(patTokens[i], 16);
                    String[] repTokens = replaceHex.split(" ");
                    byte[] rep = new byte[repTokens.length];
                    for (int i = 0; i < repTokens.length; i++) rep[i] = (byte) Integer.parseInt(repTokens[i], 16);
                    if (pat.length != rep.length) {
                        System.err.println("Длина шаблона и замены должны совпадать");
                        return;
                    }
                    int count = editor.replaceBytes(pat, rep, offset);
                    System.out.printf("Заменено: %d\n", count);
                    break;
                case "checksum":
                    if (checksumType == null) {
                        System.err.println("Укажите тип контрольной суммы: crc32, md5, sha1");
                        return;
                    }
                    switch (checksumType) {
                        case "crc32": System.out.printf("CRC32: %08X\n", editor.checksumCRC32()); break;
                        case "md5": System.out.printf("MD5: %s\n", editor.checksumMD5()); break;
                        case "sha1": System.out.printf("SHA1: %s\n", editor.checksumSHA1()); break;
                        default: System.err.println("Неизвестный тип");
                    }
                    break;
                default:
                    System.err.println("Неизвестная команда");
            }
            if (output != null) {
                // Сохранение через NIO
                try (FileChannel out = FileChannel.open(Paths.get(output), StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
                    buffer.position(0);
                    out.write(buffer);
                }
                System.out.printf("Сохранено в %s\n", output);
            }
        } finally {
            editor.close();
        }
    }
}

File: HexEditor.cs
// HexEditor.cs
// Версия на C# с использованием MemoryMappedFile, Span<byte>, CRC32

using System;
using System.IO;
using System.IO.MemoryMappedFiles;
using System.Security.Cryptography;
using System.Text;
using System.Linq;
using System.Collections.Generic;

public class HexEditor : IDisposable
{
    private readonly string filename;
    private readonly int cols;
    private readonly MemoryMappedFile mmf;
    private readonly MemoryMappedViewAccessor accessor;
    private readonly long size;
    private byte[] data; // для простоты загружаем весь файл (для малых файлов)

    public HexEditor(string filename, int cols = 16)
    {
        this.filename = filename;
        this.cols = cols;
        this.size = new FileInfo(filename).Length;
        this.mmf = MemoryMappedFile.CreateFromFile(filename, FileMode.Open, null, 0, MemoryMappedFileAccess.ReadWrite);
        this.accessor = mmf.CreateViewAccessor(0, size, MemoryMappedFileAccess.ReadWrite);
        // Читаем данные в массив для простоты (для больших файлов лучше использовать потоки)
        this.data = new byte[size];
        this.accessor.ReadArray(0, data, 0, (int)size);
    }

    public void View(long offset, long length = 0)
    {
        if (length == 0) length = size - offset;
        if (offset < 0 || offset + length > size) throw new ArgumentOutOfRangeException();
        var chunk = data.Skip((int)offset).Take((int)length).ToArray();
        Dump(chunk, offset);
    }

    private void Dump(byte[] data, long baseOffset)
    {
        for (int i = 0; i < data.Length; i += cols)
        {
            int end = Math.Min(i + cols, data.Length);
            long addr = baseOffset + i;
            Console.Write($"{Colorize($"{addr:X8}", ConsoleColor.Cyan)}  ");
            // HEX
            string hex = string.Join(" ", data.Skip(i).Take(end - i).Select(b => $"{b:X2}"));
            hex = hex.PadRight(cols * 3);
            Console.Write($"{Colorize(hex, ConsoleColor.Green)}  ");
            // ASCII
            string ascii = new string(data.Skip(i).Take(end - i).Select(b => (b >= 32 && b < 127) ? (char)b : '.').ToArray());
            Console.WriteLine($"{Colorize(ascii, ConsoleColor.Blue)}");
        }
    }

    private string Colorize(string text, ConsoleColor color)
    {
        return $"\u001b[{color switch { ConsoleColor.Cyan => 96, ConsoleColor.Green => 92, ConsoleColor.Blue => 94, _ => 0 }}m{text}\u001b[0m";
    }

    public void EditByte(long offset, byte value)
    {
        if (offset < 0 || offset >= size) throw new ArgumentOutOfRangeException();
        data[offset] = value;
        accessor.Write(offset, value);
    }

    public List<long> FindBytes(byte[] pattern, long start = 0)
    {
        var positions = new List<long>();
        for (long i = start; i <= size - pattern.Length; i++)
        {
            bool match = true;
            for (int j = 0; j < pattern.Length; j++)
            {
                if (data[i + j] != pattern[j]) { match = false; break; }
            }
            if (match) positions.Add(i);
        }
        return positions;
    }

    public int ReplaceBytes(byte[] pattern, byte[] replacement, long start = 0)
    {
        if (pattern.Length != replacement.Length) throw new ArgumentException("Длина шаблона и замены должны совпадать");
        var positions = FindBytes(pattern, start);
        foreach (var p in positions)
        {
            for (int j = 0; j < replacement.Length; j++)
            {
                data[p + j] = replacement[j];
                accessor.Write(p + j, replacement[j]);
            }
        }
        return positions.Count;
    }

    public uint ChecksumCRC32()
    {
        var crc = new CRC32();
        return crc.ComputeHash(data);
    }

    public string ChecksumMD5()
    {
        using var md5 = MD5.Create();
        return BitConverter.ToString(md5.ComputeHash(data)).Replace("-", "").ToLower();
    }

    public string ChecksumSHA1()
    {
        using var sha1 = SHA1.Create();
        return BitConverter.ToString(sha1.ComputeHash(data)).Replace("-", "").ToLower();
    }

    public void Save(string outputFile)
    {
        File.WriteAllBytes(outputFile, data);
    }

    public void Dispose()
    {
        accessor.Dispose();
        mmf.Dispose();
    }
}

// Простая реализация CRC32
public class CRC32
{
    private readonly uint[] table;
    public CRC32()
    {
        table = new uint[256];
        for (uint i = 0; i < 256; i++)
        {
            uint crc = i;
            for (int j = 0; j < 8; j++)
                crc = (crc & 1) != 0 ? (crc >> 1) ^ 0xEDB88320 : crc >> 1;
            table[i] = crc;
        }
    }
    public uint ComputeHash(byte[] data)
    {
        uint crc = 0xFFFFFFFF;
        foreach (byte b in data) crc = table[(crc ^ b) & 0xFF] ^ (crc >> 8);
        return ~crc;
    }
}

class Program
{
    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Использование: dotnet run -- <file> [команда] [опции]");
            return;
        }
        string filename = args[0];
        string command = "view";
        long offset = 0;
        int cols = 16;
        string bytesHex = null, replaceHex = null, checksumType = null, stringSearch = null, output = null;

        for (int i = 1; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-o": offset = long.Parse(args[++i]); break;
                case "--cols": cols = int.Parse(args[++i]); break;
                case "-b": bytesHex = args[++i]; break;
                case "--replace-with": replaceHex = args[++i]; break;
                case "-s": stringSearch = args[++i]; break;
                case "-c": checksumType = args[++i]; break;
                case "--output": output = args[++i]; break;
                default:
                    if (args[i] == "view" || args[i] == "edit" || args[i] == "find" || args[i] == "replace" || args[i] == "checksum")
                        command = args[i];
                    break;
            }
        }

        using var editor = new HexEditor(filename, cols);
        try
        {
            switch (command)
            {
                case "view": editor.View(offset); break;
                case "edit":
                    if (string.IsNullOrEmpty(bytesHex)) { Console.WriteLine("Укажите --bytes"); return; }
                    byte val = Convert.ToByte(bytesHex.Trim(), 16);
                    editor.EditByte(offset, val);
                    Console.WriteLine($"Байт по смещению {offset} изменён на {bytesHex}");
                    break;
                case "find":
                    byte[] pattern;
                    if (!string.IsNullOrEmpty(bytesHex))
                    {
                        pattern = bytesHex.Split(' ').Select(s => Convert.ToByte(s, 16)).ToArray();
                    }
                    else if (!string.IsNullOrEmpty(stringSearch))
                    {
                        pattern = Encoding.UTF8.GetBytes(stringSearch);
                    }
                    else { Console.WriteLine("Укажите --bytes или --string"); return; }
                    var positions = editor.FindBytes(pattern, offset);
                    if (positions.Count > 0)
                    {
                        Console.WriteLine($"Найдено вхождений: {positions.Count}");
                        foreach (var p in positions) Console.WriteLine($"  {p:X8}");
                    }
                    else Console.WriteLine("Не найдено");
                    break;
                case "replace":
                    if (string.IsNullOrEmpty(bytesHex) || string.IsNullOrEmpty(replaceHex))
                    { Console.WriteLine("Укажите --bytes и --replace-with"); return; }
                    var pat = bytesHex.Split(' ').Select(s => Convert.ToByte(s, 16)).ToArray();
                    var rep = replaceHex.Split(' ').Select(s => Convert.ToByte(s, 16)).ToArray();
                    if (pat.Length != rep.Length) { Console.WriteLine("Длина шаблона и замены должны совпадать"); return; }
                    int count = editor.ReplaceBytes(pat, rep, offset);
                    Console.WriteLine($"Заменено: {count}");
                    break;
                case "checksum":
                    if (string.IsNullOrEmpty(checksumType)) { Console.WriteLine("Укажите тип контрольной суммы: crc32, md5, sha1"); return; }
                    switch (checksumType.ToLower())
                    {
                        case "crc32": Console.WriteLine($"CRC32: {editor.ChecksumCRC32():X8}"); break;
                        case "md5": Console.WriteLine($"MD5: {editor.ChecksumMD5()}"); break;
                        case "sha1": Console.WriteLine($"SHA1: {editor.ChecksumSHA1()}"); break;
                        default: Console.WriteLine("Неизвестный тип"); break;
                    }
                    break;
                default: Console.WriteLine("Неизвестная команда"); break;
            }
            if (!string.IsNullOrEmpty(output))
            {
                editor.Save(output);
                Console.WriteLine($"Сохранено в {output}");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Ошибка: {ex.Message}");
        }
    }
}
